### ISSUE TITLE
Inline Assembly? Include .asm?

### ISSUE THREAD

**Issue by bangbangsheshotmedown** [odin-lang/Odin#4444]
Hello,

Is inline assembly planned before 1.0?

Odin looks interesting to me, but I need to be able to write assembly

If not planned, is there a way to load .asm files?

Thanks, and sorry if that's not the right place to ask, there was a "suggest new feature" button for the issue so I figured I'd try

**Comment by Feoramund** [comment]
I haven't ever seen any official documentation on it, but there is most definitely an `asm` token in the Odin tokenizer. I expect, given the low-level nature of the language, we'll get support for it eventually.

**Comment by flysand7** [comment]
> If not planned, is there a way to load .asm files?

Yes, there is. You can "foreign import" asm files directly, and the compiler will build them and link them to your executable.

```
foreign import lib "lib.asm"

foreign lib {
   some_proc :: proc "sysv" (a: i64, b: i64) -> i64 ---
}
```

As far as I remember the "linking" part doesn't apply to object files, so it might only work if you're building an executable or a shared object.

**Comment by gingerBill** [comment]
Inline assembly is "planned" but we are not saying what it will be precisely. This is mainly because when people say they want inline assembly, it's not obvious what exactly they really want. I understand people have used inline assembly in other compilers before and the two general approaches are to have a string literal (GCC style) or have actual parsed tokens within an `asm` block (traditional style). 

However I am still wondering if people even want actually _inline_ assembly or rather have a pseudo-procedure where its entire body is just assembly instead, but that body still understands basic aspects of the Odin type system.

**THE FOLLOWING IS PSEUDO-CODE AND NOT A PROPOSAL**

```odin
thingymabob :: asm(x: i32, y: i32) -> (z: f32, w: f32) {
    // assembly goes here
    // this isn't actually a procedure but when it is called, it will truly inline the code
}
```


There is the other aspect where any use of assembly is done, the file must be guarded with the specific architecture for it to be allowed, e.g. `foo_amd64.odin` or `#+build amd64`

**Comment by xzores** [comment]
As a note, i would like it to be able to use it for making an OS (a new journey for me)

**Comment by StudebakerGuy** [comment]
For me, my typicall uses would be more straight forward with true inline. However, I've also had to write funcitons in assembly and the psedo procdure would make those more straight forward or at least easier to read.

**Comment by laytan** [comment]
Thought I'd give some thoughts too:

I was doing some co-routine experiments and was not having fun without inline/better assembly support.
Here is that experiment: https://gist.github.com/laytan/1c9887a3c9847ed7073dd632d9114229

The biggest problems are that it uses an external assembler, this slows down compile time because it calls out to the assembler and waits for all asm files, it also means people have to install the assembler separately.

And maybe the biggest issue is that nasm does not support debug info annotations, so you are left with no debugging, the debugger has no idea what happened when a co-routine switch happens.


Another thing the current ways doesn't allow but my experiment didn't need is actually inlined assembly, these foreign imports of assembly files will have a function call to get there, this is a deal breaker for a lot of assembly use cases.

**Comment by PatrickHechler** [comment]
> However I am still wondering if people even want actually _inline_ assembly or rather have a pseudo-procedure where its entire body is just assembly instead, but that body still understands basic aspects of the Odin type system.
> 
> **THE FOLLOWING IS PSEUDO-CODE AND NOT A PROPOSAL**
> 
> thingymabob :: asm(x: i32, y: i32) -> (z: f32, w: f32) {
>     // assembly goes here
>     // this isn't actually a procedure but when it is called, it will truly inline the code
> }

I like the idea especially since odin has no preprocessor, this would allow to have `whatever-arch.odin` files that specify the assembler routines and then use them in other architecture independent files.
is it also intended to allow to specify what storage class the parameters/return values use (specific register, any register, constant, memory, ...) and what registers/input parameters will be overwritten (could this be done with variables/constants defined in the pseudo proc)?

 **THE FOLLOWING IS PSEUDO-CODE AND NOT A PROPOSAL**

```odin
thingymabob :: asm(x: i32 asm(register), y: i32) -> (z: f32, w: f32) {
    val: i32 asm(memory | register=r0) : 5 // constant 5, must not be modified by this block
    { // should {} be allowed to scope the temporary variables?
        tmp0: i32 asm(register) = x
    } // tmp0 is no longer visible
    tmp1: i32 asm(register) = y // tmp0 and tmp1 may use the same register when the compiler loads y here
    y = --- // should it be possible to "free" parameters/variables to allow later variables to use the register?
    y = y // should it be possible to "unfree" parameters/variables when the the asm blocks sometimes skips the freed block?
}
```

> There is the other aspect where any use of assembly is done, the file must be guarded with the specific architecture for it to be allowed, e.g. `foo_amd64.odin` or `#+build amd64`

I think `when` guards should also be allowed, otherwise it will be annoying to create a file for each supported architectures when only one small `asm` block is needed.

**Comment by Comodinoh** [comment]
Hey, is inline assembly still being worked on? The idea of assembly procedures sounds pretty interesting :)

**Comment by IogaMaster** [comment]
I believe ginger bill spoke about its inclusion in the 1.0 release of Odin

**Comment by gingerBill** [comment]
Odin now supports inline assembly as well as the nasm `foreign import` stuff.
